prov/gni/src/gnix_msg.c

```diff
@@ -558,6 +728,315 @@ static int __gnix_rndzv_req(void *arg)
 	return gnixu_to_fi_errno(status);
 }
 
+/* Attempt to post the iov txds */
+static int __gnix_rndzv_iov_req_post(void *arg)
+{
+	struct gnix_fab_req *req = (struct gnix_fab_req *)arg;
+	struct gnix_tx_descriptor *txd;
+	gni_return_t status;
+	struct gnix_nic *nic = req->gnix_ep->nic;
+
+	assert(nic != NULL);
+
+	GNIX_TRACE(FI_LOG_EP_DATA, "\n");
+
+	GNIX_DEBUG(FI_LOG_EP_DATA, "%s for req = %p\n",
+		   slist_empty(&req->iov_txd_slist) ?
+		   "req->iov_txd_slist is empty" :
+		   "req->iov_txd_slist is NON-empty");
+
+	COND_ACQUIRE(nic->requires_lock, &nic->lock);
+
+	while (!slist_empty(&req->iov_txd_slist)) {
+		txd = container_of(slist_remove_head(&req->iov_txd_slist),
+				   struct gnix_tx_descriptor,
+				   slist);
+
+		if (txd->gni_desc.type & GNI_POST_RDMA_GET) {
+			status = GNI_PostRdma(req->vc->gni_ep, &txd->gni_desc);
+		} else {
+			status = GNI_CtPostFma(req->vc->gni_ep, &txd->gni_desc);
+		}
+
+		if (status != GNI_RC_SUCCESS) {
+			COND_RELEASE(nic->requires_lock, &nic->lock);
+
+			/* We assume that __gnix_rndzv_req_build did it's job,
+			 * don't free this txd just put it at the end of the
+			 * slist. We'll try again next time the request is
+			 * dequeued by the progress loop.
+			 * slist uses tail ptr, so, O(1) to insert tail.
+			 *
+			 * TODO: Count tx retries?
+			 */
+			slist_insert_tail(&txd->slist, &req->iov_txd_slist);
+
+			GNIX_WARN(FI_LOG_EP_DATA, "%s failed on txd %p: with "
+				  "%s\n",
+				  txd->gni_desc.type & GNI_POST_RDMA_GET ?
+				  "GNI_PostRdma" : "GNI_CtPostFma", txd,
+				  gni_err_str[status]);
+
+			return gnixu_to_fi_errno(status);
+		}
+	}
+
+	COND_RELEASE(nic->requires_lock, &nic->lock);
+
+	return FI_SUCCESS;
+}
+
+/*
+ * Dequeued a iov work req on the remote endpoint, initiate the pull of data.
+ */
+static int __gnix_rndzv_iov_req_build(void *arg)
+{
+	int ret = FI_SUCCESS, i = 0, j = 0, use_tx_cq_blk;
+	struct gnix_fab_req *req = (struct gnix_fab_req *)arg;
+	struct gnix_fid_ep *ep = req->gnix_ep;
+	struct gnix_nic *nic = ep->nic;
+	gni_ep_handle_t gni_ep = req->vc->gni_ep;
+	struct gnix_tx_descriptor *txd = NULL;
+	size_t recv_len, get_len, ct_size, send_cnt, recv_cnt, txd_cnt;
+	uint64_t recv_ptr = 0UL;
+	/* TODO: Should this be the sender's rndzv thresh instead? */
+	size_t rndzv_thresh = ep->domain->params.msg_rendezvous_thresh;
+	gni_ct_get_post_descriptor_t *cur_ct = NULL;
+	void **next_ct = NULL;
+	uint32_t max_ct_size = nic->gni_cdm_modes & GNI_CDM_MODE_FMA_SHARED ?
+		1024*1024 : 1<<30;
```

Comment:
> Where do these constants come from?

These constants come from the UGNI documentation for `GNI_CtPostFma`.  Since I could not find a way to determine the communication domain modes via the UGNI library I added this check as well as the gni_cmd_modes to each nic to ensure that a chained transaction of Fma GETS does not exceed the maximum chained transaction size.